### DISCUSSION TITLE
Conventions to customize ID transform

### DISCUSSION THREAD

**Discussion by montyclt** [ravendb/ravendb#13889] [Ideas]
Hello! I know that I can use [OnBeforeStore and OnAfterConversionToEntity events](https://ayende.com/blog/188257-B/avoiding-non-string-identifiers-in-ravendb) to achieve this, but my problem is that [BeforeStore events is dispatching on SaveChanges call and not Store call](https://github.com/ravendb/ravendb/discussions/13888), causing that my entity doesn't have the ID until `SaveChanges` method call.

I propose new conventions to customize how ID is transformed like this:

```
using var documentStore = new DocumentStore
{
    Database = "pruebas",
    Urls = new[] { "http://localhost:8080" },
    Conventions = new DocumentConventions()
    {
        SetIdToEntity = (entity, identity, defaultSetId) =>
        {
            if (entity is AggregateRoot ag)
            {
                var id = new AggregateId(identity);
                ag.Id = id;
            }
            else
            {
                defaultSetId(entity, identity); // Use default behaviour that set it into (FindIdentityProperty convention) property of type string.
            }
        },
        GetIdFromEntity = (entity, defaultGetId) =>
        {
            if (entity is AggregateRoot ag)
            {
                return ag.Id.Value;
            }

            return defaultGetId(entity); // Use default behaviour that get it from (FindIdentityProperty convention) property of type string.
        }
    }
};
```

**Comment by ayende** [comment]
`BeforeStore` in invoked on `SaveChanges` because you can modify the document between the `Store` and the `SaveChanges`

I assume that you are familiar with this:

https://ayende.com/blog/188257-B/avoiding-non-string-identifiers-in-ravendb

Note that it would probably be easier to go the other way, define the RavenDB string Id property as `RavenDBId` property on `AggreageRoot` and then do the same work it the setter there.

**Reply by montyclt** [reply]
The idea of defining `RavenDBId` property and using the setter is working well, thank you, but I don't like it because it is an architecture leak on my domain classes.

I think that I can use a source generator (or Emit) on infrastructure layer that generates a proxy with the RavenDBId property...

**Reply by montyclt** [reply]
Hi again @ayende, I tried using this convection, but this only works when the property exists in the class that are storing, not in a base class.

```cs
using var documentStore = new DocumentStore
{
    // Urls omitted
    Conventions =
    {
        FindIdentityProperty = member => member.Name == "__RavenId"
    }
};
```

It looks like it only iterates over the properties of the class that I'm trying to store/load and not the properties of base class.

**Update**: If the `__RavenId` property is not private, then works.
**Update 2**: I open an [issue](https://github.com/ravendb/ravendb/issues/13893) about this.

**Comment by DejanMilicic** [comment]
Can you use something like this

```
public abstract class AggregateRoot
{
    public string Id { get; init; } = Guid.NewGuid().ToString(); // or whatever ID-generation logic might be
```

**Reply by montyclt** [reply]
Yes, I can do it, but all my application is working with a strongly-typed ids using a generic `Id<TEntity>` type that allow me using type-safe in method calls. I cannot change whole application to use strings.

**Comment by haludi** [comment]
Hi,

I run this test and it works fine.

class A
{
    public string __RavenId { set; get; }
}
class B : A
{

}

[Fact]
public async Task TestCase()
{
    const string ravenId = "B/123";

    using var store = GetDocumentStore(new Options
    {
        ModifyDocumentStore = s => s.Conventions =new DocumentConventions
        {
            FindIdentityProperty = member => member.Name == "__RavenId"
        }
    });
    using (var session = store.OpenAsyncSession())
    {
        var entity = new B {__RavenId = ravenId};
        await session.StoreAsync(entity);
        await session.SaveChangesAsync();
        Assert.Equal(ravenId, session.Advanced.GetDocumentId(entity));
    }
}


Is that what you meant?

Best regards,
Igal

**Reply by montyclt** [reply]
Hi Igal, my purpose is have a Id type that is not a string (is my specific case, is a generic `Id<TEntity>`, but I used `AggregateId` for simplicity in the OP), so in this forum thread I suggest adding new conventions that convert the internal RavenDB string to my custom type.

For the case you're testing is related, but is not the same, there are a problem if `A::__RavenId` is a private property, I open an issue https://github.com/ravendb/ravendb/issues/13893

**Comment by mcrio** [comment]
When using DDD, for my projects I always implemented a separate persistence model. It allows to easily define the custom ID types on the entities while having string types on the persistence model, and it allows not to have to mix domain logic with persistence. Also when implementing the read side I do not need to use the aggregates and entity classes but use the persistence model.

**Reply by montyclt** [reply]
Hi mcrio, thank you! I am using entities in my domain model, that doesn't depends on RavenDB, and a separate project that uses RavenDB, but referencing my entities to avoid rewriting the same classes changing only the Id property.

I only use entities in the OLTP part of my application, for queries I'm using projections that returns records, allowing me to get only the needed info, and avoid getting the entities with methods.

I'm thinking of using Source Generators for that, but I never used it.
